test/EngineTest.cpp

```diff
@@ -23,114 +32,259 @@ auto I = [](const auto& id) {
+  return result;
+}
 
-  a.clear();
-  b.clear();
-  res.clear();
+/*
+ * A structure containing all information needed for a normal join test. A
+ * normal join test is defined as two IdTables being joined, the resulting
+ * IdTable tested, if it is sorted after the join column, or not, and this
+ * IdTable than compared with the sampleSolution.
+ */
+struct normalJoinTest {
+  IdTable leftIdTable;
+  size_t leftJC; // Join column for the left IdTable.
+  IdTable rightIdTable;
+  size_t rightJC; // Join column for the right IdTable.
+  bool testForSorted; // Should the result of the join be tested, if it is
+                      // orderd after leftJC? In other words, ordered after the join column?
+  IdTable sampleSolution;
+
+  // The initialization list is needed, because IdTables can't be constructed
+  // without arguments.
+  normalJoinTest(const IdTable& LeftIdTable,
+      const size_t LeftJC,
+      const IdTable& RightIdTable,
+      const size_t RightJC,
+      const IdTable& SampleSolution,
+      const bool TestForSorted = false): leftIdTable(LeftIdTable.clone()),
+      leftJC(LeftJC),
+      rightIdTable(RightIdTable.clone()),
+      rightJC(RightJC),
+      testForSorted(TestForSorted),
+      sampleSolution(SampleSolution.clone())
+  {}
+};
 
-  for (size_t i = 1; i <= 10000; ++i) {
-    a.push_back({I(4 + i), I(2 + i)});
+/*
+ * @brief Tests, if the given IdTable has the same content as the sample
+ * solution and, if the option was choosen, if the IdTable is sorted by
+ * the join column.
+ *
+ * @tparam TABLE_WIDTH Just ignore it. Made to be infered from the function
+ *  paramters.
+ *
+ * @param table The IdTable that should be tested.
+ * @param sampleSolution The sample solution. Doesn't need to be sorted,
+ *  or the same order of rows as the table.
+ * @param testForSorted If this is true, it will also be tested, if the table
+ *  is sorted by the join column.
+ * @param jc The join column of the table.
+ * @param t Ignore it. It's only here for being able to make better messages,
+ *  if a IdTable fails the comparison.
+*/
+void compareIdTableWithSolution(const IdTable& table, 
+    const IdTable& sampleSolution,
+    const bool testForSorted = false,
+    const size_t jc = 0,
+    ad_utility::source_location t = ad_utility::source_location::current()
+    ) {
+  // For generating more informative messages, when failing the comparison.
+  std::stringstream traceMessage;
+  traceMessage << "compareIdTableWithSolution comparing IdTable\n";
+  for (size_t row = 0; row < table.size(); row++) {
+    for (size_t column = 0; column < table.numColumns(); column++) {
+      traceMessage << table(row, column) << " ";
+    }
+    traceMessage << "\n";
+  }
+  traceMessage << "with IdTable \n";
+  for (size_t row = 0; row < sampleSolution.size(); row++) {
+    for (size_t column = 0; column < sampleSolution.numColumns(); column++) {
+      traceMessage << sampleSolution(row, column) << " ";
+    }
+    traceMessage << "\n";
+  }
```

Comment:
This code is duplicated. It should be factored out into a lambda, or into a free function `writeIdTableToString` or `...toStream`, depending on how you implement it.

Reply:
I partly already did that, but I improved it a bit in [7ab3290](https://github.com/ad-freiburg/qlever/pull/821/commits/7ab329051f54effff2da3077ed2faef9ed39cd5f).

Reply:
Never mind, had an unnecessary named lambda in that commit. Got rid of that in [4e6c77e](https://github.com/ad-freiburg/qlever/pull/821/commits/4e6c77e2dc8f35404c7ad8bf036b77b11ca53b82).